<?xml version="1.0" encoding="UTF-8"?>
<launch>

  <arg name="headless"/>
  <arg name="world_model_file" default="worlds/empty.world"/>
  <arg name="robot_gt_urdf_file"/>
  <arg name="robot_realistic_urdf_file"/>
  <arg name="output" default="screen"/>

  <param name="/use_sim_time" value="true"/>

  <arg name="paused" default="false"/>
  <arg name="extra_gazebo_args" default=""/>
  <arg name="recording" default="false"/>
  <arg name="debug" default="false"/>
  <arg name="physics" default="ode"/>
  <arg name="verbose" default="true"/>
  <arg name="respawn_gazebo" default="false"/>
  <arg name="use_clock_frequency" default="false"/>
  <arg name="pub_clock_frequency" default="100"/>
  <arg name="enable_ros_network" default="true" />
  <arg name="server_required" default="true"/>
  <arg name="gui_required" default="false"/>

  <!-- set command arguments -->
  <arg unless="$(arg paused)" name="command_arg1" value=""/>
  <arg     if="$(arg paused)" name="command_arg1" value="-u"/>
  <arg unless="$(arg recording)" name="command_arg2" value=""/>
  <arg     if="$(arg recording)" name="command_arg2" value="-r"/>
  <arg unless="$(arg verbose)" name="command_arg3" value=""/>
  <arg     if="$(arg verbose)" name="command_arg3" value="--verbose"/>
  <arg unless="$(arg debug)" name="script_type" value="gzserver"/>
  <arg     if="$(arg debug)" name="script_type" value="debug"/>

  <!-- start gazebo server-->
  <group if="$(arg use_clock_frequency)">
    <param name="gazebo/pub_clock_frequency" value="$(arg pub_clock_frequency)" />
  </group>
  <group>
    <param name="gazebo/enable_ros_network" value="$(arg enable_ros_network)" />
  </group>

  <node name="gazebo"
        pkg="gazebo_ros"
        type="$(arg script_type)"
        respawn="$(arg respawn_gazebo)"
        output="$(arg output)"
        args="$(arg command_arg1) $(arg command_arg2) $(arg command_arg3) -e $(arg physics) $(arg extra_gazebo_args) $(arg world_model_file)"
        required="$(arg server_required)" />

  <node name="gazebo_gui"
        pkg="gazebo_ros"
        type="gzclient"
        respawn="false"
        output="$(arg output)"
        args="$(arg command_arg3)"
        required="$(arg gui_required)"
        unless="$(arg headless)"/>

  <param name="robot_gt_urdf" textfile="$(arg robot_gt_urdf_file)"/>
  <node name="robot_state_publisher_gt"
        pkg="robot_state_publisher"
        type="robot_state_publisher">
    <remap from="robot_description" to="robot_gt_urdf" />
    <remap from="joint_states" to="joint_states_gt" />
  </node>

  <param name="robot_realistic_urdf" textfile="$(arg robot_realistic_urdf_file)"/>
  <node name="robot_state_publisher_realistic"
        pkg="robot_state_publisher"
        type="robot_state_publisher">
    <remap from="robot_description" to="robot_realistic_urdf" />
    <remap from="joint_states" to="joint_states_realistic" />
  </node>

  <node name="gt_odom_static_transform_publisher"
        pkg="tf2_ros"
        type="static_transform_publisher"
        output="$(arg output)"
        args="0 0 0  0 0 0  map odom"/>

</launch>
